contributors/devel/sig-api-machinery/controllers.md

```diff
@@ -93,98 +95,102 @@ func NewController(pods informers.PodInformer) *Controller {
 	pods.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
-			key, err := cache.MetaNamespaceKeyFunc(obj)
+			ref, err := cache.ObjectToName(obj)
 			if err == nil {
-				c.queue.Add(key)
+				c.queue.Add(ref)
 			}
 		},
 		UpdateFunc: func(old interface{}, new interface{}) {
-			key, err := cache.MetaNamespaceKeyFunc(new)
+			ref, err := cache.ObjectToName(new)
 			if err == nil {
-				c.queue.Add(key)
+				c.queue.Add(ref)
 			}
 		},
 		DeleteFunc: func(obj interface{}) {
 			// IndexerInformer uses a delta nodeQueue, therefore for deletes we have to use this
 			// key function.
-			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			ref, err := cache.DeletionHandlingObjectToName(obj)
 			if err == nil {
-				c.queue.Add(key)
+				c.queue.Add(ref)
 			}
 		},
 	},)
 	
 	return c
 }
 
-func (c *Controller) Run(threadiness int, stopCh chan struct{}) {
+func (c *Controller) Run(ctx context.Context, threadiness int) error {
 	// don't let panics crash the process
 	defer utilruntime.HandleCrash()
 	// make sure the work queue is shutdown which will trigger workers to end
 	defer c.queue.ShutDown()
+	logger := klog.FromContext(ctx)
 
-	klog.Infof("Starting <NAME> controller")
+	logger.Info("Starting <NAME> controller")
 
 	// wait for your secondary caches to fill before starting your work
-	if !cache.WaitForCacheSync(stopCh, c.podsSynced) {
-		return
+	if !cache.WaitForCacheSync(ctx.Done(), c.podsSynced) {
+		return fmt.Errorf("failed to wait for caches to sync")
 	}
 
 	// start up your worker threads based on threadiness.  Some controllers
 	// have multiple kinds of workers
 	for i := 0; i < threadiness; i++ {
 		// runWorker will loop until "something bad" happens.  The .Until will
 		// then rekick the worker after one second
-		go wait.Until(c.runWorker, time.Second, stopCh)
+		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
 	}
+	logger.Info("Started workers")
 
 	// wait until we're told to stop
-	<-stopCh
-	klog.Infof("Shutting down <NAME> controller")
+	<-ctx.Done()
+	logger.Info("Shutting down <NAME> controller")
+
+	return nil
 }
 
-func (c *Controller) runWorker() {
+func (c *Controller) runWorker(ctx context.Context) {
 	// hot loop until we're told to stop.  processNextWorkItem will
 	// automatically wait until there's work available, so we don't worry
 	// about secondary waits
-	for c.processNextWorkItem() {
+	for c.processNextWorkItem(ctx) {
 	}
 }
 
 // processNextWorkItem deals with one key off the queue.  It returns false
 // when it's time to quit.
-func (c *Controller) processNextWorkItem() bool {
-	// pull the next work item from queue.  It should be a key we use to lookup
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// pull the next work item from queue.  It should be an object reference that we use to lookup
 	// something in a cache
-	key, quit := c.queue.Get()
-	if quit {
+	item, shutdown := c.queue.Get()
```

Comment:
```suggestion
	ref, shutdown := c.queue.Get()
```

and throughout for consistency with how its queued.

Reply:
Done